### PR TITLE
WIP - feat(libraries): add section about libraries

### PR DESCRIPTION
**Pull request by emirose**
## Overview

Adding docs for what shared libraries might look like to follow up on https://github.com/telusdigital/architecture-forum/issues/86.  I've only included docs for the pieces that everyone seemed to agree on. There are several open questions still about how we want to approach ownership, discoverability, and maintenance.

Do we have guidelines for how to approach open sourcing something? If so, we might want to link to that from here.

I'm also not sure where this section makes the most sense. I've put it under development for now.

Feedback please! @theetrain @ryanoglesby08 @petermarshall878 @wascloud 

### Features

- create new page to describe approach to sharing libraries and components

---

#### Meta

- [x] provide a descriptive topic
- [x] provide an overview of contribution
- [x] no sensitive content included, such as:
  - security & privacy policy violating content
  - content considered competitive intelligence
  - keys, tokens or credentials
- [x] documentation format follows [this template][template]
- [x] fork is up to date ([see this guide from github](https://help.github.com/articles/syncing-a-fork/))
- [x] "work in progress" commits are squashed ([see "Squashing Commits"](https://git-scm.com/book/id/v2/Git-Tools-Rewriting-History))
- [x] commits follow the [Karma][karma-format] or [Commitizen](https://www.npmjs.com/package/commitizen) format

[template]: ../.template.md
[karma-format]: https://karma-runner.github.io/1.0/dev/git-commit-msg.html

### CODE REVIEW

File: development/libraries.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+## What
+
+Teams should consider creating an npm library for reuse across TELUS when:
+
+* There's Telus specific logic that we want to share between applications
+* There's no existing open source alternative for the problem we're trying to solve
+* We're confident that the solution is relatively stable
+* Library responsibilities are clear
+
+If the logic that we want to reuse is not TELUS specific, we should consider publishing the package publicly.
```
Comment (theetrain) [inline development/libraries.md]: I understand it's unclear when something should be 'open source'. By default employee obligations, all npm packages must be private by default and then published by the `@telusdigital` npm account. On a case-by-case basis, your immediate manager can allow public releases of packages if the use case is broad enough for non-TELUS use. It's also possible to self-publish a package if it is usable for non-TELUS consumers **and** you have spent a large part of your personal time developing it.

File: development/libraries.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+* Be driven by the needs of the consumer
+* Be decoupled from the implementation of the consuming application
+* Have easy to consume APIs
+* Be explicit about dependencies
+* Include documentation with example usage
+* Follow the versioning strategy defined [here](versioning.md)
+* Include automated tests
+* Have clear contribution guidelines
+
+Teams should consider using the [npm library starter Kit](https://github.com/telusdigital/npm-library-starter-kit) as a starting point.
```
Comment (ryanoglesby08) [inline development/libraries.md]: Should this be more "forceful?" Maybe it should be less of "consider this" and more of "you should use this". That could drive more adoption of it and force it to get more mature.

File: development/libraries.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+Any library should:
+
+* Be driven by the needs of the consumer
+* Be decoupled from the implementation of the consuming application
+* Have easy to consume APIs
+* Be explicit about dependencies
+* Include documentation with example usage
+* Follow the versioning strategy defined [here](versioning.md)
+* Include automated tests
+* Have clear contribution guidelines
```
Comment (ryanoglesby08) [inline development/libraries.md]: In general I like these points. They feel a bit light though. There is so much more so say about all of this. Is the intention to be concise? Do you think more detail is needed to help explain what is meant or give recommendations about how to do these things well?

File: development/libraries.md (hunk 1 above)
Comment (crimeminister) [reply development/libraries.md]: Is there a legal review that has to happen for code to be open sourced? If nothing else, I imagine we'd need an approval around the license used.

File: development/libraries.md (hunk 3 above)
Comment (crimeminister) [reply development/libraries.md]: I'd be interested in seeing a description of when _not_ to create a library, e.g. in the context of a service architecture an argument can be made that factoring into libraries is an anti-pattern.

@ryanoglesby08 Would you be alright with merging _as is_, and fleshing these points out in successive PRs?

File: development/libraries.md (hunk 2 above)
Comment (crimeminister) [reply development/libraries.md]: +1 on changing the language to be more forceful.

**Comment by crimeminister** [comment]
@ryanoglesby08, @theetrain: I will assume this PR since AFAIK we have lost @emirose.

File: development/libraries.md (hunk 2 above)
Comment (crimeminister) [reply development/libraries.md]: More specifically, how about changing the wording to, "Teams SHOULD consider using..."; the capitalization is important as I'd like to move to RFC 2119 verbiage to indicate conformance requirements.

Comment (ryanoglesby08) [reply development/libraries.md]: I suppose. I don't know much about the RFC 2119 verbiage so I'll defer to your judgement.

However, why not just removing the words "should" and/or "consider"

"Teams must use the npm library starter kit."
"Use the npm library starter kit."

File: development/libraries.md (hunk 3 above)
Comment (ryanoglesby08) [reply development/libraries.md]: Yeah sure go for it.

**Comment by emirose** [comment]
Sorry for the slow response on this. I'll make some updates based on the above feedback. To the question about whether the points are too vague/high-level I'm open to ideas on what to make more specific. For the first pass, I was trying to capture the ideas that everyone seemed to agree on, which all ended up being a bit vague :-)

File: development/libraries.md (hunk 1 above)
Comment (emirose) [reply development/libraries.md]: Maybe this part should be taken out then? Or documented elsewhere? There were two themes that came up when this was originally discussed: 1) we don't want to create TELUS libraries when there are existing open source packages that do the same thing (or something very similar) and 2) there might be some good opportunities for TELUS to contribute back to the community around the tools we use. Does this belong in here or in a separate article?

Comment (crimeminister) [reply development/libraries.md]: Linking to a detailed discussion of the criteria for publishing as open source to another document seems reasonable. We could request that a review process be undertaken to provide some confidence that we're not overlapping with an existing product, encourage upstream contributions to the packages we use, discuss licensing, etc.

**Comment by crimeminister** [comment]
Thanks for revisiting this @emirose. I think it'll be more productive to entertain specific proposals in this area in future PRs. I'm inclined to merge this after you've made your update(s) based on the already-provided feedback.

File: development/libraries.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+## Why
+
+To prevent teams from duplicating work and to allow for easier maintenance of our applications, teams may want to create [npm libraries](npm.md) to allow for reuse of components or logic. Defining standards and principals behind this type of code reuse will help us to reduce the effort required by teams to use and maintain these libraries.
+
+## What
+
+Teams should consider creating an npm library for reuse across TELUS when:
+
+* There's TELUS specific logic that we want to share between applications
```
Comment (theetrain) [inline development/libraries.md]: `TELUS-specific`

File: development/libraries.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+* We're confident that the solution is relatively stable
+* Library responsibilities are clear
+
+## How
+
+Any library should:
+
+* Be driven by the needs of the consumer
+* Be decoupled from the implementation of the consuming application
+* Have easy to consume APIs
```
Comment (theetrain) [inline development/libraries.md]: `easy-to-consume`

**Pull request merged**